src/devtools/client/inspector/rules/components/RulesApp.tsx

```diff
@@ -0,0 +1,195 @@
+import { Rule } from "devtools/client/inspector/rules/components/Rule";
+import { Rules } from "devtools/client/inspector/rules/components/Rules";
+import { Toolbar } from "devtools/client/inspector/rules/components/Toolbar";
+
+import { getStr } from "devtools/client/inspector/rules/utils/l10n";
+import { useSelector } from "react-redux";
+import { RulesState, RuleState } from "../state/rules";
+
+const SHOW_PSEUDO_ELEMENTS_PREF = "devtools.inspector.show_pseudo_elements";
+
+type RulesAppProps = {
+  onToggleDeclaration: Function;
+  onToggleSelectorHighlighter: Function;
+  rules: RuleState[];
+  showContextMenu: Function;
+  showDeclarationNameEditor: Function;
+  showDeclarationValueEditor: Function;
+  showNewDeclarationEditor: Function;
+  showSelectorEditor: Function;
+};
+
+export const RulesApp: FC<RulesAppProps> = ({
+  showContextMenu,
+  onToggleDeclaration,
+  showDeclarationValueEditor,
+  showNewDeclarationEditor,
+  onToggleSelectorHighlighter,
+  showDeclarationNameEditor,
+  showSelectorEditor,
+}) => {
+  const { rules } = useSelector((state: { rules: RulesState }) => ({
+    rules: state.rules.rules || [],
+  }));
+
+  const [rulesQuery, setRulesQuery] = useState("");
+
+  const ruleProps = {
+    onToggleDeclaration,
+    onToggleSelectorHighlighter,
+    showDeclarationNameEditor,
+    showDeclarationValueEditor,
+    showNewDeclarationEditor,
+    showSelectorEditor,
+    query: rulesQuery,
+  };
+
+  const onContextMenu = (event: React.MouseEvent<HTMLDivElement>) => {
+    if (
+      event.currentTarget.closest("input[type=text]") ||
+      event.currentTarget.closest("input:not([type])") ||
+      event.currentTarget.closest("textarea")
+    ) {
+      return;
+    }
+
+    event.stopPropagation();
+    event.preventDefault();
+
+    showContextMenu(event);
+  };
+
+  const renderInheritedRules = (rules: RuleState[]) => {
+    if (!rules.length) {
+      return null;
+    }
+
+    const output = [];
+    let lastInheritedNodeId;
+
+    for (const rule of rules) {
+      if (!rule.inheritance) {
+        continue;
+      }
+
+      const { inheritedNodeId, inheritedSource } = rule.inheritance;
+
+      if (inheritedNodeId !== lastInheritedNodeId) {
+        lastInheritedNodeId = inheritedNodeId;
+        output.push(
+          <div key={inheritedNodeId} className="ruleview-header">
+            {inheritedSource}
+          </div>
+        );
+      }
+
+      output.push(
+        <Rule key={`${inheritedNodeId}|${rule.id}`} rule={rule} {...ruleProps} query={rulesQuery} />
+      );
+    }
+
+    return output;
+  };
+
+  const renderStyleRules = (rules: RuleState[]) => {
+    if (!rules.length) {
+      return null;
+    }
+
+    return <Rules {...ruleProps} rules={rules} />;
+  };
+
+  const renderPseudoElementRules = (rules: RuleState[]) => {
+    if (!rules.length) {
+      return null;
+    }
+
+    type FCProps<C> = C extends FC<infer P> ? P : never;
+    const componentProps: FCProps<typeof Rules> = {
```

Comment:
Nice!

Reply:
Yeah, I didn't want to waste time migrating an entire Accordion, but I wanted to make sure I didn't forget something, so here we are... :D